### PR TITLE
Add support for Decimal object to Double cast in ML Flow 

### PR DESCRIPTION
**Pull request by shitaoli-db**
## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?
Added unit test for Decimal to float object in enforce schema cast.
Also added end to end test with databricks automl model inference, which has double schema in model, the inference was successful

image
image

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

Model inference now allow user to use Decimal object as input. Note decimal input will only be allowed when schema is double and the data will be casted to double and thus may lost precision.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pyfunc/__init__.py [hunk 1]
```diff
@@ -397,6 +402,15 @@ def _enforce_mlflow_datatype(name, values: pandas.Series, t: DataType):
             raise MlflowException(
                 "Failed to convert column {0} from type {1} to {2}.".format(name, values.dtype, t)
             )
+    if t == DataType.double and values.dtype == decimal.Decimal:
```
Comment (yxiong) [inline mlflow/pyfunc/__init__.py:405]: I thought in our test the `values.dtype` is `object`, while `type(values[0])` is `decimal.Decimal`. Can you confirm?

**Review by yxiong — CHANGES_REQUESTED** [review]

File: mlflow/pyfunc/__init__.py (hunk 1 above)
Comment (shitaoli-db) [reply mlflow/pyfunc/__init__.py:405]: yeah, verified
image

Comment (shitaoli-db) [reply mlflow/pyfunc/__init__.py:405]: Also our unit test and e2e notebook test all passed when object is Decimal

Comment (yxiong) [reply mlflow/pyfunc/__init__.py:405]: Ah, I didn't know `dtype == object` and `dtype == decimal.Decimal` can both be true. LGTM!

**Review by yxiong — APPROVED** [review]

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -397,6 +402,15 @@ def _enforce_mlflow_datatype(name, values: pandas.Series, t: DataType):
...
+        # NB: Pyspark Decimal columne get converted to decimal.Decimal when converted to pandas
```
Comment (tomasatdatabricks) [inline mlflow/pyfunc/__init__.py:406]: nit: columne -> column.
in order to support ...  - is part of the comment missing?

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good.

Can you also update docs please?
Can we make this work only in spark_udf environment or protect it with an argument? E.g. we could use an env variable or extra argument to pyfunc.load_model. Would that be feasible?

File: mlflow/pyfunc/__init__.py (hunk 2 above)
Comment (shitaoli-db) [reply mlflow/pyfunc/__init__.py]: Thank you. Done.

**Comment by shitaoli-db** [comment]
> 

Let me try with an extra argument first. Do you know how did we generate this message?

image
If we use the extra argument, then we have to customize this message in the UI to let customer know that if they want their model with decimal training data work in inference, they have to pass an extra argument.


> Looks good.
> 
> Can you also update docs please? Can we make this work only in spark_udf environment or protect it with an argument? E.g. we could use an env variable or extra argument to pyfunc.load_model. Would that be feasible?

File: mlflow/pyfunc/__init__.py [hunk 3]
```diff
@@ -709,7 +729,10 @@ def _warn_dependency_requirement_mismatches(model_path):
 
 
 def load_model(
-    model_uri: str, suppress_warnings: bool = False, dst_path: str = None
+    model_uri: str,
+    suppress_warnings: bool = False,
+    dst_path: str = None,
+    support_decimal: bool = False,
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py:735]: Is there a way to accomplish the goal of providing conditional decimal enforcement without adding this flag to the `mlflow.pyfunc.load_model()` API? I'd rather not extend this API for a niche use case.

Comment (yxiong) [reply mlflow/pyfunc/__init__.py:735]: I personally don't think we need this flag at all. Can we just always enable this conversion? I don't see any real use case of setting that to false (or only add that flag later when we confirm it's a real use case).

@dbczumar @tomasatdatabricks what do you think?

Comment (shitaoli-db) [reply mlflow/pyfunc/__init__.py:735]: I think tom provided 2 approach to me 
image

The other option I could think on top of my head is use environment variable

https://github.com/mlflow/mlflow/blob/master/mlflow/environment_variables.py

I am not sure which approach is more preferrable, let's have a quick agreement.

Comment (dbczumar) [reply mlflow/pyfunc/__init__.py:735]: I agree with @yxiong here

Comment (shitaoli-db) [reply mlflow/pyfunc/__init__.py]: Reverted back the change, now we always enable the cast.

**Review by yxiong — APPROVED** [review]

**Review by dbczumar — APPROVED** [review]
LGTM!

**Comment by shitaoli-db** [comment]
autoformat

**Comment by shitaoli-db** [comment]
automerge

**Pull request merged**